core/src/main/java/io/grpc/internal/PickFirstLeafLoadBalancer.java

```diff
@@ -465,9 +465,13 @@ public void onSubchannelState(ConnectivityStateInfo newState) {
       log.log(Level.FINE, "Received health status {0} for subchannel {1}",
           new Object[]{newState, subchannelData.subchannel});
       healthStateInfo = newState;
-      if (addressIndex.isValid()
-          && subchannels.get(addressIndex.getCurrentAddress()).healthListener == this) {
-        updateHealthCheckedState(subchannelData);
+      try {
+        SubchannelData curSubChanData = subchannels.get(addressIndex.getCurrentAddress());
+        if (curSubChanData != null && curSubChanData.healthListener == this) {
+          updateHealthCheckedState(subchannelData);
+        }
+      } catch (IllegalStateException e) {
```

Comment:
I don't think we should be catching an IllegalStateException. And this handling logs but throws away the exception.

Reply:
The IllegalStateException comes from calling addressIndex.getCurrentAddress when the index is off the end of the list of addresses.  Since this isn't synchronized, the addressIndex.isValid check isn't completely guaranteed to avoid the situation so I got rid of the isValid check and just catch the exception if thrown.  Since all of this logic is just to determine whether the listener was triggered for the currently active subchannel, we really don't care which of the various reasons that is not the case.  Logging was so that if we have to debug we get a better sense of what is going on.

Reply:
> Since this isn't synchronized

It is synchronized; it is run on the synchronization context. If it isn't synchronized, all bets would be off and we could see data corruption/tears.